catenets/models/torch/base.py

```diff
@@ -0,0 +1,514 @@
+            for b in range(n_batches):
+                self.optimizer.zero_grad()
+
+                idx_next = train_indices[
+                    (b * batch_size) : min((b + 1) * batch_size, n - 1)
+                ]
+
+                X_next = X[idx_next]
+                y_next = y[idx_next].squeeze()
+
+                preds = self.forward(X_next).squeeze()
+
+                batch_loss = self.loss(preds, y_next)
+
+                batch_loss.backward()
+
+                self.optimizer.step()
+                train_loss.append(batch_loss.detach())
+
+            train_loss = torch.Tensor(train_loss).to(DEVICE)
+
+            if i % self.n_iter_print == 0:
+                with torch.no_grad():
+                    preds = self.forward(X_val).squeeze()
+                    val_loss = self.loss(preds, y_val)
+                    if val_loss_best > val_loss:
+                        val_loss_best = val_loss
+                        patience = 0
+                    else:
+                        patience += 1
+                    if patience > self.patience and i > self.n_iter_min:
+                        break
+                    log.info(
+                        f"[{self.name}] Epoch: {i}, current {val_string} loss: {val_loss}, train_loss: {torch.mean(train_loss)}"
+                    )
+
+        return self
+
+    def _check_tensor(self, X: torch.Tensor) -> torch.Tensor:
+        if isinstance(X, torch.Tensor):
+            return X.to(DEVICE)
+        else:
+            return torch.from_numpy(np.asarray(X)).to(DEVICE)
+
+
+class BaseCATEEstimator(nn.Module):
+    """
+    Interface for estimators of CATE.
+
+    The interface has train/forward API for PyTorch-based models and fit/predict API for sklearn-based models.
+    """
+
+    def __init__(
+        self,
+    ) -> None:
+        super(BaseCATEEstimator, self).__init__()
+
+    @abc.abstractmethod
+    @check_input_train
+    @benchmark
+    def train(
+        self,
+        X: torch.Tensor,
+        y: torch.Tensor,
+        w: torch.Tensor,
+    ) -> "BaseCATEEstimator":
+        """
+        Train method for a CATEModel
+
+        Parameters
+        ----------
+        X: torch.Tensor
+            Covariate matrix
+        y: torch.Tensor
+            Outcome vector
+        w: torch.Tensor
+            Treatment indicator
+        """
+        ...
+
+    @benchmark
+    def fit(
+        self,
+        X: torch.Tensor,
+        y: torch.Tensor,
+        w: torch.Tensor,
+    ) -> "BaseCATEEstimator":
+        return self.train(X, y, w)
+
+    @abc.abstractmethod
+    @benchmark
+    def forward(self, X: torch.Tensor) -> torch.Tensor:
+        """
+        Predict treatment effect estimates using a CATE estimator.
+
+        Parameters
+        ----------
+        X: pd.DataFrame or np.array
+            Covariate matrix
+
+        Returns
+        -------
+        potential outcomes probabilities
+        """
+        ...
+
+    @benchmark
+    def predict(

```

Comment:
Does something speak against keeping the return_po flag, so that its possible to return both treatment effect and potential outcomes if desired?

Reply:
It is just a bit weird to not have a standard format for the return values, for any parameters.

But I can add the return_po logic as well, sure.